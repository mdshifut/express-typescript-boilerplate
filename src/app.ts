import dotenv from "dotenv";
dotenv.config(); // Initialized  env variables
import config from "./startup/config";
config(); // Initialized config options

import express, { Application } from "express";
import logger from "./logger/logger";
// tslint:disable-next-line: no-var-requires
require("express-async-errors");
import db from "./startup/db";
import middleware from "./startup/middleware";
import routes from "./startup/routes";

const app: Application = express(); // Initialized  express app

db(); // Initialized  database
middleware(app); // Initialized  middleware
routes(app); // Initialized  routes

const port = process.env.PORT || 5000;
const server = app.listen(port, () =>
  logger.info(`Listening on port ${port}......`)
);

export default server;
