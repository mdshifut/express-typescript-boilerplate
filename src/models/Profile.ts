import { Document, model, Schema } from "mongoose";

export interface IProfile extends Document {
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
}

const profileSchema: Schema = new Schema({
  firstName: { type: String, required: true },
  lastName: { type: String, required: true },
  email: { type: String, required: true },
  phone: { type: String, required: true }
});

export default model<IProfile>("Profile", profileSchema);
