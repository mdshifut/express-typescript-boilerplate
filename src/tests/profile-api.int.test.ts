import _ from "lodash";
import request, { Response } from "supertest";

import app from "../app";
import Profile from "../models/Profile";

interface IProfile {
  firstName?: string;
  lastName?: string;
  email?: string;
  phone?: string;
}

const testProfile: IProfile = {
  firstName: "John",
  lastName: "Doe",
  email: "test@email.com",
  phone: "896 7454 6546"
};

describe("Profile route => api/profile", () => {
  beforeAll(async () => {
    await Profile.deleteMany({});
  });

  afterAll(async () => {
    app.close();
  });

  describe("POST /create-profile", () => {
    const exec = (value: IProfile): Promise<Response> => {
      return request(app)
        .post("/api/profile/create-profile")
        .send(value);
    };

    it("should return error if data is not provide", async () => {
      const { status } = await exec({});

      expect(status).toBe(400);
    });

    /**
     * ================ firstName ==================
     */
    describe("firstName", () => {
      it("should return error if firstName is not provided", async () => {
        const { status, body } = await exec(_.omit(testProfile, ["firstName"]));

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/firstName/);
      });

      it("should return error if firstName length less than 3 characters", async () => {
        const { status, body } = await exec({
          ...testProfile,
          firstName: "te"
        });

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/firstName/);
        expect(body.error.message).toMatch(/3/);
      });

      it("should return error if firstName length greater than 20 characters", async () => {
        const { status, body } = await exec({
          ...testProfile,
          firstName: "testgreaterthantwientycharceters"
        });

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/firstName/);
        expect(body.error.message).toMatch(/20/);
      });
    }); // firstName

    /**
     * ================ lastName ==================
     */
    describe("lastName", () => {
      it("should return error if lastName is not provided", async () => {
        const { status, body } = await exec(_.omit(testProfile, ["lastName"]));

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/lastName/);
      });

      it("should return error if lastName length less than 3 characters", async () => {
        const { status, body } = await exec({
          ...testProfile,
          lastName: "te"
        });

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/lastName/);
        expect(body.error.message).toMatch(/3/);
      });

      it("should return error if lastName length greater than 20 characters", async () => {
        const { status, body } = await exec({
          ...testProfile,
          lastName: "testgreaterthantwientycharceters"
        });

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/lastName/);
        expect(body.error.message).toMatch(/20/);
      });
    }); // lastName

    /**
     * ================ email ==================
     */
    describe("email", () => {
      it("should return error if email is not provided", async () => {
        const { status, body } = await exec(_.omit(testProfile, ["email"]));

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/email/);
      });

      it("should return error if email is not a valid email", async () => {
        const { status, body } = await exec({
          ...testProfile,
          email: "invalidEmail"
        });

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/email/);
        expect(body.error.message).toMatch(/invalid/);
      });
    }); // email

    /**
     * ================ phone ==================
     */
    describe("phone", () => {
      it("should return error if phone is not provided", async () => {
        const { status, body } = await exec(_.omit(testProfile, ["phone"]));

        expect(status).toBe(400);
        expect(body.error.message).toMatch(/phone/);
      });
    }); // phone

    it("should create profile and return success message if all data is valid", async () => {
      const { status, body } = await exec(testProfile);

      const profile = await Profile.findOne({ email: testProfile.email });

      expect(status).toBe(201);
      expect(body.message).toMatch(/successfully/);
      expect(profile).toMatchObject(testProfile);
    });
  });
});
