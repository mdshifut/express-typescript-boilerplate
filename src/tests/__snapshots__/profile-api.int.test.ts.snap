// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile route => api/profile POST /create-profile should create profile and return success message if all data is valid 1`] = `
Object {
  "$__": InternalCache {
    "$options": Object {
      "isNew": false,
      "skipId": true,
      "willInit": true,
    },
    "$setCalled": Set {},
    "_id": "5e2ee3d526e28e0f737a77b5",
    "activePaths": StateMachine {
      "paths": Object {
        "__v": "init",
        "_id": "init",
        "email": "init",
        "firstName": "init",
        "lastName": "init",
        "phone": "init",
      },
      "stateNames": Array [
        "require",
        "modify",
        "init",
        "default",
        "ignore",
      ],
      "states": Object {
        "default": Object {},
        "ignore": Object {},
        "init": Object {
          "__v": true,
          "_id": true,
          "email": true,
          "firstName": true,
          "lastName": true,
          "phone": true,
        },
        "modify": Object {},
        "require": Object {},
      },
    },
    "adhocPaths": undefined,
    "cachedRequired": Object {},
    "emitter": EventEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": 0,
    },
    "fullPath": undefined,
    "getters": Object {},
    "inserting": undefined,
    "ownerDocument": undefined,
    "pathsToScopes": Object {},
    "populate": undefined,
    "populated": undefined,
    "removing": undefined,
    "saveError": undefined,
    "saving": undefined,
    "scope": undefined,
    "selected": Object {},
    "session": undefined,
    "shardval": undefined,
    "strictMode": true,
    "validationError": undefined,
    "version": undefined,
    "wasPopulated": false,
  },
  "$init": true,
  "$locals": Object {},
  "_doc": Object {
    "__v": 0,
    "_id": "5e2ee3d526e28e0f737a77b5",
    "email": "test@email.com",
    "firstName": "John",
    "lastName": "Doe",
    "phone": "896 7454 6546",
  },
  "email": "test@email.com",
  "errors": undefined,
  "firstName": "John",
  "isNew": false,
  "lastName": "Doe",
  "phone": "896 7454 6546",
}
`;
